test/common/upstream/eds_speed_test.cc

```diff
@@ -0,0 +1,202 @@
+// Note: this should be run with --compilation_mode=opt, and would benefit from a
+// quiescent system with disabled cstate power management.
+
+#include "envoy/config/cluster/v3/cluster.pb.h"
+#include "envoy/config/core/v3/health_check.pb.h"
+#include "envoy/config/endpoint/v3/endpoint.pb.h"
+#include "envoy/config/endpoint/v3/endpoint_components.pb.h"
+#include "envoy/service/discovery/v3/discovery.pb.h"
+#include "envoy/stats/scope.h"
+
+#include "common/config/utility.h"
+#include "common/singleton/manager_impl.h"
+#include "common/upstream/eds.h"
+
+#include "server/transport_socket_config_impl.h"
+
+#include "test/common/upstream/utility.h"
+#include "test/mocks/local_info/mocks.h"
+#include "test/mocks/protobuf/mocks.h"
+#include "test/mocks/runtime/mocks.h"
+#include "test/mocks/server/mocks.h"
+#include "test/mocks/ssl/mocks.h"
+#include "test/mocks/upstream/mocks.h"
+#include "test/test_common/utility.h"
+
+#include "benchmark/benchmark.h"
+
+namespace Envoy {
+namespace Upstream {
+
+class EdsSpeedTest {
+public:
+  EdsSpeedTest() : api_(Api::createApiForTest(stats_)) { resetCluster(); }
+
+  void resetCluster() {
+    resetCluster(R"EOF(
+      name: name
+      connect_timeout: 0.25s
+      type: EDS
+      lb_policy: ROUND_ROBIN
+      eds_cluster_config:
+        service_name: fare
+        eds_config:
+          api_config_source:
+            api_type: REST
+            cluster_names:
+            - eds
+            refresh_delay: 1s
+    )EOF",
+                 Cluster::InitializePhase::Secondary);
+  }
+
+  void resetClusterDrainOnHostRemoval() {
```

Comment:
This seems pretty complicated for what you'd expect in a basic EDS performance benchmark to validate this change, how come the need for this kind of specialization?

Reply:
There are competing goals of providing a lightweight speed test and measuring updates that look something like production EDS updates. That said I haven't given it much detailed consideration and can take a simplification pass. This test is mostly eds_test with large datasets wrapped in a benchmark test.

Reply:
I think from the perspective of what you want to achieve in this test, it would be better to simplify, since this is a microbenchmark.

Reply:
I've streamlined this a bunch, but there's still some duplication between it and eds_test.